### DISCUSSION TITLE
Creating an extension to add flowing columns to Word & ODT

### DISCUSSION THREAD

**Discussion by jimjam-slam** [quarto-dev/quarto-cli#1541] [Show and Tell]
This is maybe more of a Pandoc question/discussion than a Quarto one, but I'll throw it in here anyway!

I've been exploring the potential to create a Quarto extension implementing a Pandoc filter that turns fenced divs with a class on them into flowing column sections.

The `.columns` class for HTML formats in Quarto creates explicit columns (where you explicitly identify the content that goes in each column), but it's common to have columns in many formats where content overflowing the first column flows into the next one.

This is easy enough in HTML with a fenced div and `style="columns: 2"`, `style="columns: 6rem auto"`, etc. But in OpenOffice ODT and Word documents, [it's a real pain in the bum](https://github.com/jimjam-slam/columnflow/issues/1).

In Word, sections do not have separate parent elements; instead, you add the style information (eg. column spec) to the final paragraph of a section, unless it's the final section, in which case you add it to the end of the body. (This is still blowing my mind.) So theoretically, you need to walk the document to identify all potential sections, add the "previous section" before your fenced div, then work out whether your fenced one is the final one and place the style info appropriately.

In OpenOffice it at least makes more sense, but it's potentially even less possible for a filter: you tag the section content using a `text:section` parent element, then you add a style with the column spec to `styles.xml`, and you link the two with a `style-name` attribute. I'm pretty sure this is impossible to achieve in a filter without providing an appropriately structured reference document that already has a section style in it.

Maybe the filter can load the reference document and modify its contents raw though? Perhaps a Quarto extension could bundle its own reference document in - seems like a shame when most people will want to bring their own though! (I would argue that Pandoc's reliance on reference docs for style information is a weakness of the format generally, but 🤷🏻‍♂️)

Any ideas or brainstorming here would be appreciated! I've set up a skeleton extension, but so far it's, uhh, very skeletal 😆  https://github.com/jimjam-slam/columnflow

**Comment by jimjam-slam** [comment]
I could be wrong about this, but I also believe that Word's style editor doesn't include the column options, which makes it difficult to bring in a reference document that includes support for extra columns (and also makes sense, since section break info is inlined into the paragraph).

I'm less familiar with OpenOffice, but I believe [its style editor _does_ allow you to configure section styles that include columns](https://wiki.openoffice.org/wiki/Documentation/UserGuide/Writer/Styles/Page#Columns_Tab) (which also makes sense since column info is stored as section style info in `.odt` files).

**Reply by jimjam-slam** [reply]
So I am making some progress with this! Proof-of-concept extension for adding flowing columns to Word docs: https://github.com/jimjam-slam/columnflow

Hoping to add essential features later in the week and then look at ODT support! That's trickier, because (afaict) OFT keeps column info in style definitions, which are in a separate XML that comes from the reference doc. Not sure if there's a way to in-line it in the doc content yet.

**Comment by mcanouil** [comment]
I believe there is no need for an extension:

```
---
title: "mysite"
format: docx
---

This is a Quarto website.

To learn more about Quarto websites visit <https://quarto.org/docs/websites>.


::: {layout-ncol=2}
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc finibus varius venenatis. Vivamus tincidunt, ex vitae sollicitudin ultrices, lectus massa sagittis eros, id interdum erat augue eget lacus. Vivamus nec nibh pulvinar, volutpat odio a, auctor arcu. Donec ac rutrum leo. Aliquam in diam pulvinar, aliquet mauris quis, laoreet lorem. Duis ut feugiat augue, sit amet tincidunt erat. Integer vitae arcu tristique, dapibus nibh at, rhoncus urna. Proin tincidunt massa eu ligula scelerisque feugiat. Vivamus et maximus tortor. Proin suscipit interdum erat in bibendum. Nulla facilisi. Duis commodo sapien aliquam erat pulvinar, a semper turpis porttitor. Nulla facilisi.

Praesent quis ex odio. Integer a interdum nisl, a luctus orci. Aenean tincidunt ornare ligula vitae dapibus. Pellentesque vitae mauris consectetur mi aliquam efficitur. Nullam scelerisque turpis ante, eu bibendum quam commodo id. Mauris sagittis gravida arcu mollis congue. Etiam gravida tellus vestibulum turpis pretium, in eleifend ipsum euismod. Praesent vel velit eu dolor suscipit elementum. Morbi feugiat augue sapien, pretium venenatis ex venenatis egestas. Maecenas eget mi velit. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque ornare sollicitudin quam, a consequat est sodales at. Mauris porttitor turpis aliquam lacinia faucibus. Integer at nisi egestas dolor hendrerit mollis.

Vivamus pretium auctor luctus. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc sodales sit amet quam a gravida. Praesent vestibulum felis at arcu blandit, a imperdiet turpis sodales. Sed elementum, ligula sit amet efficitur tempus, mi risus dictum erat, vel sollicitudin magna dolor et metus. Fusce volutpat orci lectus, a porttitor nisl semper eu. Fusce viverra enim et tellus tristique, sit amet ultrices odio efficitur. Mauris laoreet gravida nisi nec faucibus.

Sed elit nulla, lobortis et metus quis, porttitor venenatis ligula. Vestibulum laoreet feugiat lacus in lobortis. Nam magna magna, laoreet in massa a, egestas porttitor risus. Morbi dolor nulla, auctor eget dolor ut, porta efficitur arcu. Sed a aliquam diam, eget malesuada nisi. Aenean facilisis, magna ac cursus pretium, massa mauris fermentum leo, id gravida lorem leo nec massa. Vestibulum ultricies ac nunc et eleifend. Ut faucibus tortor condimentum dolor vehicula porta ac sed ligula. In ornare purus non est imperdiet vulputate. Morbi et magna ut augue lobortis gravida. Praesent aliquam vulputate massa, ut dignissim mi convallis sed. Vestibulum in laoreet augue, quis dapibus tortor. Mauris velit odio, condimentum ut metus nec, consectetur mattis elit. Phasellus ac lacus accumsan, mollis eros vel, efficitur urna.

Mauris sit amet lacus id metus tempus venenatis. Pellentesque eget ullamcorper arcu. Ut at nibh turpis. Cras non quam a lectus aliquet tincidunt. Cras quis mi ligula. Sed luctus risus nibh, non tempus lacus sollicitudin ac. Aenean auctor feugiat ligula eget luctus. Suspendisse ac ipsum ut magna finibus condimentum vitae id odio. Fusce lobortis, dolor vitae aliquet consequat, eros leo ullamcorper diam, quis laoreet libero ipsum at turpis.
:::

```

<img width="485" alt="Screenshot of a page in a Word document showing a two columns layout filled with a lorem ipsum text" src="https://user-images.githubusercontent.com/8896044/181926114-992ef2d2-3981-412a-9e5c-c18cb8031d3e.png">

**Reply by jimjam-slam** [reply]
Is this flowing content between columns, or is it allocating blocks (pars) to columns based on their count?

Flowing columns are nice because (a) they can break paragraphs across columns where required, and (b) if someone subsequently adds or removes content to the output document, the surrounding content reflows to the correct column.

**Reply by mcanouil** [reply]
The code is provided, as you can see it's neither a or b, and also both a and b.

**Reply by jimjam-slam** [reply]
I'm not sure I understand what you mean, but I'll have a play and see how it works!

**Reply by mcanouil** [reply]
Try my example, you'll see.
It's a whole paragraphs overflow, meaning if a paragraph does not fit in the column, it's added in the next one. I don't set manually what's inside the columns contrary to what the following will require:

```
:::: {.columns}

::: {.column width="40%"}
contents...
:::

::: {.column width="60%"}
contents...
:::

::::
```

**Reply by jimjam-slam** [reply]
Mmm, yeah, it's definitely different to using `.columns`, where the author assigns content. I guess the problem is that the allocation of content to columns, while handled by Quarto (or Pandoc), is still baked in (and handled at block level) at render time.

So here's your above example:

![image](https://user-images.githubusercontent.com/6520659/182049842-8f5d8c50-cbba-40de-924d-f7af24ced4fb.png)

If I now open that Word document and insert a par (here in red text) after par 2, you can see it doesn't correctly flow to the next column - instead it flows to the next page:

![image](https://user-images.githubusercontent.com/6520659/182049937-2a843d96-6ea8-4e13-ad8a-8ab1c301d739.png)

Or, if I delete par 2, par 4 becomes a full-width section instead of moving up to take par 3's place while par 3 takes par 2's place:

![image](https://user-images.githubusercontent.com/6520659/182049976-051939ba-ad17-4e48-9ae9-e9d688082530.png)

In contrast, if I make a doc with a two-column section, it looks like this:

![image](https://user-images.githubusercontent.com/6520659/182050034-f9413359-2cd9-41a6-bfc8-45097e1c35b4.png)

I can insert text (here in red):

![image](https://user-images.githubusercontent.com/6520659/182050062-d66d202c-c8fd-4101-be98-3b7d8d009ace.png)

Or remove par 2:

![image](https://user-images.githubusercontent.com/6520659/182050083-28a87f79-b96e-4c3f-aae6-7ae75d9e3eec.png)

I appreciate the thought, and `layout-ncol` is definitely useful in a lot of situations! I just don't think it's quite what I'm looking for in this case 🙂

**Comment by jimjam-slam** [comment]
The basics of this are working now! Still gotta shore it up a bit with nested fenced divs and such, but it's [looking good](https://github.com/jimjam-slam/columnflow).

**Comment by jimjam-slam** [comment]
Working on [a PR to improve the insertion of column specs into the document](https://github.com/jimjam-slam/columnflow/pull/4), but it's not ready yet! Still getting weird results in a lot of cases 🤔
